File: problems/032-conflict-package-cover/solutions/python/main.py
import sys, collections

a = list(map(int, sys.stdin.read().split()))
L, R, M = a[:3]
g = [[] for _ in range(L)]
p = 3
for _ in range(M):
    g[a[p] - 1].append(a[p + 1] - 1)
    p += 2
pu = [-1] * L
pv = [-1] * R
dist = [-1] * L
matching = 0
while True:
    q = collections.deque()
    for u in range(L):
        dist[u] = 0 if pu[u] < 0 else -1
        if pu[u] < 0:
            q.append(u)
    terminal = -1
    while q:
        u = q.popleft()
        if terminal >= 0 and dist[u] >= terminal:
            continue
        for v in g[u]:
            w = pv[v]
            if w < 0:
                terminal = dist[u]
            elif dist[w] < 0:
                dist[w] = dist[u] + 1
                q.append(w)
    if terminal < 0:
        break
    cur = [0] * L
    for root in range(L):
        if pu[root] >= 0:
            continue
        su = [root]
        sv = []
        ok = False
        while su:
            u = su[-1]
            while cur[u] < len(g[u]):
                v = g[u][cur[u]]
                cur[u] += 1
                w = pv[v]
                if w < 0 and dist[u] == terminal:
                    pu[u] = v
                    pv[v] = u
                    for i in range(len(sv) - 1, -1, -1):
                        pu[su[i]] = sv[i]
                        pv[sv[i]] = su[i]
                    ok = True
                    break
                if w >= 0 and dist[u] < terminal and dist[w] == dist[u] + 1:
                    sv.append(v)
                    su.append(w)
                    break
            if ok:
                break
            if su and cur[su[-1]] >= len(g[su[-1]]):
                dist[su[-1]] = -1
                su.pop()
                if sv:
                    sv.pop()
        if ok:
            matching += 1
print(matching)

File: problems/032-conflict-package-cover/solutions/rust/main.rs
use std::collections::VecDeque;
use std::io::{self, Read};
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut it = s.split_whitespace();
    let l: usize = it.next().unwrap().parse().unwrap();
    let r: usize = it.next().unwrap().parse().unwrap();
    let m: usize = it.next().unwrap().parse().unwrap();
    let mut g = vec![vec![]; l];
    for _ in 0..m {
        let u = it.next().unwrap().parse::<usize>().unwrap() - 1;
        let v = it.next().unwrap().parse::<usize>().unwrap() - 1;
        g[u].push(v);
    }
    let mut pu = vec![usize::MAX; l];
    let mut pv = vec![usize::MAX; r];
    let mut dist = vec![-1i32; l];
    let mut matching = 0;
    loop {
        let mut q = VecDeque::new();
        for u in 0..l {
            dist[u] = if pu[u] == usize::MAX {
                q.push_back(u);
                0
            } else {
                -1
            }
        }
        let mut terminal = -1;
        while let Some(u) = q.pop_front() {
            if terminal >= 0 && dist[u] >= terminal {
                continue;
            }
            for &v in &g[u] {
                let w = pv[v];
                if w == usize::MAX {
                    terminal = dist[u]
                } else if dist[w] < 0 {
                    dist[w] = dist[u] + 1;
                    q.push_back(w)
                }
            }
        }
        if terminal < 0 {
            break;
        }
        let mut cur = vec![0; l];
        for root in 0..l {
            if pu[root] != usize::MAX {
                continue;
            }
            let (mut su, mut sv) = (vec![root], vec![]);
            let mut ok = false;
            while !su.is_empty() && !ok {
                let u = *su.last().unwrap();
                let mut down = false;
                while cur[u] < g[u].len() {
                    let v = g[u][cur[u]];
                    cur[u] += 1;
                    let w = pv[v];
                    if w == usize::MAX && dist[u] == terminal {
                        pu[u] = v;
                        pv[v] = u;
                        for i in (0..sv.len()).rev() {
                            pu[su[i]] = sv[i];
                            pv[sv[i]] = su[i]
                        }
                        ok = true;
                        break;
                    }
                    if w != usize::MAX && dist[u] < terminal && dist[w] == dist[u] + 1 {
                        sv.push(v);
                        su.push(w);
                        down = true;
                        break;
                    }
                }
                if !ok && !down {
                    dist[u] = -1;
                    su.pop();
                    sv.pop();
                }
            }
            if ok {
                matching += 1
            }
        }
    }
    println!("{matching}");
}
